flex/engines/http_server/handler/graph_db_http_handler.cc

```diff
@@ -13,109 +13,634 @@
+            outer_span->End();
+#endif  // HAVE_OPENTELEMETRY_CPP
           }
-          auto result = fut.get0();
-          rep->write_body("bin", std::move(result.content));
           rep->done();
           return seastar::make_ready_future<
               std::unique_ptr<seastar::httpd::reply>>(std::move(rep));
-        });
+        }
+        auto result = fut.get0();
+        rep->write_body("bin", std::move(result.content));
+#ifdef HAVE_OPENTELEMETRY_CPP
+        std::map<std::string, std::string> labels = {{"status", "success"}};
+        total_counter_->Add(1, labels);
+        outer_span->End();
+        auto end_ts = gs::GetCurrentTimeStamp();
+#if OPENTELEMETRY_ABI_VERSION_NO >= 2
+        latency_histogram_->Record(end_ts - start_ts);
+#else
+        latency_histogram_->Record(end_ts - start_ts,
+                                   opentelemetry::context::Context{});
+#endif
+#endif  // HAVE_OPENTELEMETRY_CPP
+        rep->done();
+        return seastar::make_ready_future<
+            std::unique_ptr<seastar::httpd::reply>>(std::move(rep));
+      });
+}
+#endif  // BUILD_HQPS
+
+///////////////////////////graph_db_http_handler/////////////////////////////
+
+graph_db_http_handler::graph_db_http_handler(uint16_t http_port,
+                                             int32_t shard_num,
+                                             bool enable_hqps_handlers)
+    : http_port_(http_port),
+      enable_hqps_handlers_(enable_hqps_handlers),
+      running_(false),
+      actors_running_(true) {
+  graph_db_handlers_.resize(shard_num);
+#ifdef BUILD_HQPS
+  if (enable_hqps_handlers_) {
+    ic_handlers_.resize(shard_num);
+    adhoc_query_handlers_.resize(shard_num);
   }
+#endif  // BUILD_HQPS
+}
 
- private:
-  const uint32_t shard_concurrency_;
-  query_dispatcher dispatcher_;
-  std::vector<executor_ref> executor_refs_;
-};
+graph_db_http_handler::~graph_db_http_handler() {
+  if (is_running()) {
+    stop();
+  }
+  // DO NOT DELETE the handler pointers, they will be deleted by
+  // seastar::httpd::match_rule
+}
 
-class graph_db_exit_handler : public seastar::httpd::handler_base {
- public:
-  seastar::future<std::unique_ptr<seastar::httpd::reply>> handle(
-      const seastar::sstring& path,
-      std::unique_ptr<seastar::httpd::request> req,
-      std::unique_ptr<seastar::httpd::reply> rep) override {
-    GraphDBService::get().set_exit_state();
-    rep->write_body("bin",
-                    seastar::sstring{"The graph_db server is exiting ..."});
-    return seastar::make_ready_future<std::unique_ptr<seastar::httpd::reply>>(
-        std::move(rep));
+uint16_t graph_db_http_handler::get_port() const { return http_port_; }
+
+bool graph_db_http_handler::is_running() const { return running_.load(); }
+
+bool graph_db_http_handler::is_actors_running() const {
+  return actors_running_.load();
+}
+
+seastar::future<> graph_db_http_handler::stop_query_actors() {
```

Comment:
in `stop_query_actors()` and `start_query_actors()`, we don't manipulate `graph_db_handler`_`, since these two functions will only be called when hqps_service is launched.